test/tools/cmap_spec_runner.ts

```diff
@@ -0,0 +1,338 @@
+
+  constructor(hostAddress) {
+    this.threads = new Map();
+    this.connections = new Map();
+    this.orphans = new Set();
+    this.poolEvents = [];
+    this.poolEventsEventEmitter = new EventEmitter();
+    this.hostAddress = hostAddress;
+    this.supportedOperations = getTestOpDefinitions(this);
+  }
+
+  getThread(name) {
+    let thread = this.threads.get(name);
+    if (!thread) {
+      thread = new Thread(this.supportedOperations);
+      this.threads.set(name, thread);
+    }
+
+    return thread;
+  }
+
+  createPool(options) {
+    this.pool = new ConnectionPool({ ...options, hostAddress: this.hostAddress });
+    ALL_POOL_EVENTS.forEach(ev => {
+      this.pool.on(ev, x => {
+        this.poolEvents.push(x);
+        this.poolEventsEventEmitter.emit('poolEvent');
+      });
+    });
+  }
+
+  closePool() {
+    return new Promise(resolve => {
+      ALL_POOL_EVENTS.forEach(ev => this.pool.removeAllListeners(ev));
+      this.pool.close(resolve);
+    });
+  }
+
+  async tearDown() {
+    if (this.pool) {
+      await this.closePool();
+    }
+    const connectionsToDestroy = Array.from(this.orphans).concat(
+      Array.from(this.connections.values())
+    );
+    const promises = connectionsToDestroy.map(conn => {
+      return new Promise<void>((resolve, reject) =>
+        conn.destroy({ force: true }, err => {
+          if (err) return reject(err);
+          resolve();
+        })
+      );
+    });
+    await Promise.all(promises);
+    this.poolEventsEventEmitter.removeAllListeners();
+  }
+}
+
+export async function runCmapTest(test: CmapTest, threadContext: ThreadContext) {
+  expect(knownTestKeys).to.include.members(Object.keys(test));
+
+  const poolOptions = test.poolOptions || {};
+  const operations = test.operations;
+  const expectedError = test.error;
+  const expectedEvents = test.events || [];
+  const ignoreEvents = test.ignore || [];
+
+  let actualError;
+
+  const MAIN_THREAD_KEY = Symbol('Main Thread');
+  const mainThread = threadContext.getThread(MAIN_THREAD_KEY);
+  mainThread.start();
+
+  threadContext.createPool(poolOptions);
+
+  for (const idx in operations) {
+    const op = operations[idx];
+
+    const threadKey = op.name === 'checkOut' ? op.thread || MAIN_THREAD_KEY : MAIN_THREAD_KEY;
+    const thread = threadContext.getThread(threadKey);
+
+    thread.queue(op);
+  }
+
+  await mainThread.finish().catch(e => {
+    actualError = e;
+  });
+
+  if (expectedError) {
+    expect(actualError).to.exist;
+    const { type: errorType, ...errorPropsToCheck } = expectedError;
+    expect(actualError).to.have.property('name', `Mongo${errorType}`);
+    compareInputToSpec(actualError, errorPropsToCheck);
+  } else {
+    expect(actualError).to.not.exist;
+  }
+
+  const actualEvents = threadContext.poolEvents.filter(
+    ev => !ignoreEvents.includes(getEventType(ev))
+  );
+
+  expect(actualEvents).to.have.lengthOf(expectedEvents.length);
+  for (const expected of expectedEvents) {
```

Comment:
```suggestion
for (const [expected, actual] of zip(expectedEvents, actualEvents)) {
    const { type: eventType, ...eventPropsToCheck } = expected;
    expect(actual.constructor.name).to.equal(`${eventType}Event`);
    compareInputToSpec(actual, eventPropsToCheck);
}
```

Reply:
what's the benefit beyond personal preference?

Reply:
It's just preference - I don't like mutating objects when I can help it.  Either way is okay

Reply:
I don't think I would consider this a preference as long as there is reason to *not* mutate a value. A reason I can think of is that when you are debugging, you cannot go back and reference what was in the array on an earlier iteration because it is gone from the list. But I do think this is just QoL, but QoL is important too 🙂  (but not a blocker)

Reply:
In this case, actualEvents exists specifically for this check, so I don't think the extra dependency import gains us anything